File: packages/tui/src/components/select-list.ts
import { EventEmitter } from "node:events";
import { getKeybindings } from "../keybindings.ts";
import type { Component } from "../tui.ts";
import { truncateToWidth, visibleWidth } from "../utils.ts";

const DEFAULT_PRIMARY_COLUMN_WIDTH = 32;
const PRIMARY_COLUMN_GAP = 2;
const MIN_DESCRIPTION_WIDTH = 10;

/**
 * Module-scope event bus for SelectList lifecycle, intended for remote-control
 * extensions that want to mirror selectors to a non-TUI client (mobile app, web).
 *
 * Events:
 *   "mount"   ({ id, items })  emitted from the SelectList constructor.
 *   "dismiss" ({ id })         emitted exactly once when the list is resolved
 *                              (selection, cancel, or remote response).
 *
 * Listeners receive the same `id` they can use with `respondToSelectList(id, value)`
 * to drive the selection remotely.
 */
export const selectListEvents = new EventEmitter();

// Registry of live SelectLists by id so a remote client can find one and drive
// its selection. Held by WeakRef (with a FinalizationRegistry) so a list the
// owner drops without an explicit confirm/cancel — there is no Component
// teardown hook to clean up on — can still be garbage-collected and its entry
// reclaimed instead of leaking. Settling a list also removes its entry eagerly.
const liveSelectLists = new Map<string, WeakRef<SelectList>>();
const selectListFinalizer = new FinalizationRegistry<string>((id) => {
	liveSelectLists.delete(id);
});

let __selectListIdCounter = 0;
function nextSelectListId(): string {
	__selectListIdCounter++;
	return `sl_${Date.now().toString(36)}_${__selectListIdCounter.toString(36)}`;
}

/**
 * Resolve a live SelectList from outside the TUI focus stack by id.
 * Pass `value === null` to cancel; otherwise the value is matched against
 * `SelectItem.value` (preferred) or `SelectItem.label` (fallback) before
 * invoking the list's onSelect callback.
 *
 * @returns for a cancel, true if a live list existed; for a selection, true
 *   only if `value` matched an item. false if no live list exists for `id`.
 */
export function respondToSelectList(id: string, value: string | null): boolean {
	const list = liveSelectLists.get(id)?.deref();
	if (!list) return false;
	if (value === null) {
		list.cancelRemote();
		return true;
	}
	return list.selectRemote(value);
}

const normalizeToSingleLine = (text: string): string => text.replace(/[\r\n]+/g, " ").trim();
const clamp = (value: number, min: number, max: number): number => Math.max(min, Math.min(value, max));

export interface SelectItem {
	value: string;
	label: string;
	description?: string;
}

export interface SelectListTheme {
	selectedPrefix: (text: string) => string;
	selectedText: (text: string) => string;
	description: (text: string) => string;
	scrollInfo: (text: string) => string;
	noMatch: (text: string) => string;
}

export interface SelectListTruncatePrimaryContext {
	text: string;
	maxWidth: number;
	columnWidth: number;
	item: SelectItem;
	isSelected: boolean;
}

export interface SelectListLayoutOptions {
	minPrimaryColumnWidth?: number;
	maxPrimaryColumnWidth?: number;
	truncatePrimary?: (context: SelectListTruncatePrimaryContext) => string;
}

export class SelectList implements Component {
	/** Stable id for this instance. See `selectListEvents` for remote control. */
	public readonly id: string = nextSelectListId();
	private items: SelectItem[] = [];
	private filteredItems: SelectItem[] = [];
	private selectedIndex: number = 0;
	private maxVisible: number = 5;
	private theme: SelectListTheme;
	private layout: SelectListLayoutOptions;
	private dismissed: boolean = false;

	public onSelect?: (item: SelectItem) => void;
	public onCancel?: () => void;
	public onSelectionChange?: (item: SelectItem) => void;

	constructor(items: SelectItem[], maxVisible: number, theme: SelectListTheme, layout: SelectListLayoutOptions = {}) {
		this.items = items;
		this.filteredItems = items;
		this.maxVisible = maxVisible;
		this.theme = theme;
		this.layout = layout;
		liveSelectLists.set(this.id, new WeakRef(this));
		selectListFinalizer.register(this, this.id, this);
		selectListEvents.emit("mount", { id: this.id, items });
	}

	/**
	 * Transition to the dismissed state exactly once. Returns true if this call
	 * performed the transition (the caller "won the race"), false if the list was
	 * already dismissed. Callers must only fire onSelect/onCancel when this
	 * returns true, so a keyboard confirm and a concurrent remote response can't
	 * both invoke the callback.
	 */
	private settle(): boolean {
		if (this.dismissed) return false;
		this.dismissed = true;
		liveSelectLists.delete(this.id);
		selectListFinalizer.unregister(this);
		selectListEvents.emit("dismiss", { id: this.id });
		return true;
	}

	/**
	 * Drive a selection from outside the TUI (e.g., a phone client).
	 * Matches by `value` first, then `label`.
	 * @returns true if `value` matched an item, false otherwise.
	 */
	selectRemote(value: string): boolean {
		const item = this.items.find((i) => i.value === value) ?? this.items.find((i) => i.label === value);
		if (!item) return false;
		if (this.settle()) this.onSelect?.(item);
		return true;
	}

	/** Drive a cancel from outside the TUI. Fires onCancel if not already dismissed. */
	cancelRemote(): void {
		if (this.settle()) this.onCancel?.();
	}

	/**
	 * Component teardown hook. Called by the TUI's overlay manager when this
	 * list is permanently removed (hide / pop). Settles the list if it hasn't
	 * been already, so `dismiss` fires deterministically on every close path —
	 * not just user pick/cancel and remote response. Does not invoke
	 * onSelect/onCancel: the unmount itself isn't a user choice.
	 */
	dispose(): void {
		this.settle();
	}

	setFilter(filter: string): void {
		this.filteredItems = this.items.filter((item) => item.value.toLowerCase().startsWith(filter.toLowerCase()));
		// Reset selection when filter changes
		this.selectedIndex = 0;
	}

	setSelectedIndex(index: number): void {
		this.selectedIndex = Math.max(0, Math.min(index, this.filteredItems.length - 1));
	}

	invalidate(): void {
		// No cached state to invalidate currently
	}

	render(width: number): string[] {
		const lines: string[] = [];

		// If no items match filter, show message
		if (this.filteredItems.length === 0) {
			lines.push(this.theme.noMatch("  No matching commands"));
			return lines;
		}

		const primaryColumnWidth = this.getPrimaryColumnWidth();

		// Calculate visible range with scrolling
		const startIndex = Math.max(
			0,
			Math.min(this.selectedIndex - Math.floor(this.maxVisible / 2), this.filteredItems.length - this.maxVisible),
		);
		const endIndex = Math.min(startIndex + this.maxVisible, this.filteredItems.length);

		// Render visible items
		for (let i = startIndex; i < endIndex; i++) {
			const item = this.filteredItems[i];
			if (!item) continue;

			const isSelected = i === this.selectedIndex;
			const descriptionSingleLine = item.description ? normalizeToSingleLine(item.description) : undefined;
			lines.push(this.renderItem(item, isSelected, width, descriptionSingleLine, primaryColumnWidth));
		}

		// Add scroll indicators if needed
		if (startIndex > 0 || endIndex < this.filteredItems.length) {
			const scrollText = `  (${this.selectedIndex + 1}/${this.filteredItems.length})`;
			// Truncate if too long for terminal
			lines.push(this.theme.scrollInfo(truncateToWidth(scrollText, width - 2, "")));
		}

		return lines;
	}

	handleInput(keyData: string): void {
		const kb = getKeybindings();
		// Up arrow - wrap to bottom when at top
		if (kb.matches(keyData, "tui.select.up")) {
			this.selectedIndex = this.selectedIndex === 0 ? this.filteredItems.length - 1 : this.selectedIndex - 1;
			this.notifySelectionChange();
		}
		// Down arrow - wrap to top when at bottom
		else if (kb.matches(keyData, "tui.select.down")) {
			this.selectedIndex = this.selectedIndex === this.filteredItems.length - 1 ? 0 : this.selectedIndex + 1;
			this.notifySelectionChange();
		}
		// Enter
		else if (kb.matches(keyData, "tui.select.confirm")) {
			const selectedItem = this.filteredItems[this.selectedIndex];
			if (selectedItem && this.settle()) {
				this.onSelect?.(selectedItem);
			}
		}
		// Escape or Ctrl+C
		else if (kb.matches(keyData, "tui.select.cancel")) {
			if (this.settle()) this.onCancel?.();
		}
	}

	private renderItem(
		item: SelectItem,
		isSelected: boolean,
		width: number,
		descriptionSingleLine: string | undefined,
		primaryColumnWidth: number,
	): string {
		const prefix = isSelected ? "→ " : "  ";
		const prefixWidth = visibleWidth(prefix);

		if (descriptionSingleLine && width > 40) {
			const effectivePrimaryColumnWidth = Math.max(1, Math.min(primaryColumnWidth, width - prefixWidth - 4));
			const maxPrimaryWidth = Math.max(1, effectivePrimaryColumnWidth - PRIMARY_COLUMN_GAP);
			const truncatedValue = this.truncatePrimary(item, isSelected, maxPrimaryWidth, effectivePrimaryColumnWidth);
			const truncatedValueWidth = visibleWidth(truncatedValue);
			const spacing = " ".repeat(Math.max(1, effectivePrimaryColumnWidth - truncatedValueWidth));
			const descriptionStart = prefixWidth + truncatedValueWidth + spacing.length;
			const remainingWidth = width - descriptionStart - 2; // -2 for safety

			if (remainingWidth > MIN_DESCRIPTION_WIDTH) {
				const truncatedDesc = truncateToWidth(descriptionSingleLine, remainingWidth, "");
				if (isSelected) {
					return this.theme.selectedText(`${prefix}${truncatedValue}${spacing}${truncatedDesc}`);
				}

				const descText = this.theme.description(spacing + truncatedDesc);
				return prefix + truncatedValue + descText;
			}
		}

		const maxWidth = width - prefixWidth - 2;
		const truncatedValue = this.truncatePrimary(item, isSelected, maxWidth, maxWidth);
		if (isSelected) {
			return this.theme.selectedText(`${prefix}${truncatedValue}`);
		}

		return prefix + truncatedValue;
	}

	private getPrimaryColumnWidth(): number {
		const { min, max } = this.getPrimaryColumnBounds();
		const widestPrimary = this.filteredItems.reduce((widest, item) => {
			return Math.max(widest, visibleWidth(this.getDisplayValue(item)) + PRIMARY_COLUMN_GAP);
		}, 0);

		return clamp(widestPrimary, min, max);
	}

	private getPrimaryColumnBounds(): { min: number; max: number } {
		const rawMin =
			this.layout.minPrimaryColumnWidth ?? this.layout.maxPrimaryColumnWidth ?? DEFAULT_PRIMARY_COLUMN_WIDTH;
		const rawMax =
			this.layout.maxPrimaryColumnWidth ?? this.layout.minPrimaryColumnWidth ?? DEFAULT_PRIMARY_COLUMN_WIDTH;

		return {
			min: Math.max(1, Math.min(rawMin, rawMax)),
			max: Math.max(1, Math.max(rawMin, rawMax)),
		};
	}

	private truncatePrimary(item: SelectItem, isSelected: boolean, maxWidth: number, columnWidth: number): string {
		const displayValue = this.getDisplayValue(item);
		const truncatedValue = this.layout.truncatePrimary
			? this.layout.truncatePrimary({
					text: displayValue,
					maxWidth,
					columnWidth,
					item,
					isSelected,
				})
			: truncateToWidth(displayValue, maxWidth, "");

		return truncateToWidth(truncatedValue, maxWidth, "");
	}

	private getDisplayValue(item: SelectItem): string {
		return item.label || item.value;
	}

	private notifySelectionChange(): void {
		const selectedItem = this.filteredItems[this.selectedIndex];
		if (selectedItem && this.onSelectionChange) {
			this.onSelectionChange(selectedItem);
		}
	}

	getSelectedItem(): SelectItem | null {
		const item = this.filteredItems[this.selectedIndex];
		return item || null;
	}
}

File: python/src/pi_mono/tui/components/select_list.py
"""SelectList component - list selection with filtering and keyboard navigation"""

from typing import List, Optional, Callable
from dataclasses import dataclass

from pi_mono.tui.utils import truncate_to_width, visible_width
from pi_mono.tui.keybindings import get_keybindings
from pi_mono.tui.editor_component import Component

DEFAULT_PRIMARY_COLUMN_WIDTH = 32
PRIMARY_COLUMN_GAP = 2
MIN_DESCRIPTION_WIDTH = 10


def _normalize_to_single_line(text: str) -> str:
    return text.replace("\r\n", " ").replace("\r", " ").replace("\n", " ").strip()


def _clamp(value: int, min_val: int, max_val: int) -> int:
    return max(min_val, min(value, max_val))


@dataclass
class SelectItem:
    value: str
    label: str
    description: Optional[str] = None


@dataclass
class SelectListTheme:
    selected_prefix: Callable[[str], str]
    selected_text: Callable[[str], str]
    description: Callable[[str], str]
    scroll_info: Callable[[str], str]
    no_match: Callable[[str], str]


@dataclass
class SelectListLayoutOptions:
    min_primary_column_width: Optional[int] = None
    max_primary_column_width: Optional[int] = None
    truncate_primary: Optional[Callable] = None


class SelectList(Component):
    """SelectList component - list selection with filtering and keyboard navigation"""

    def __init__(
        self,
        items: List[SelectItem],
        max_visible: int,
        theme: SelectListTheme,
        layout: Optional[SelectListLayoutOptions] = None,
    ) -> None:
        self._items = items
        self._filtered_items = list(items)
        self._selected_index = 0
        self._max_visible = max_visible
        self._theme = theme
        self._layout = layout or SelectListLayoutOptions()

        self.on_select: Optional[Callable[[SelectItem], None]] = None
        self.on_cancel: Optional[Callable[[], None]] = None
        self.on_selection_change: Optional[Callable[[SelectItem], None]] = None

    def set_filter(self, filter_text: str) -> None:
        if not filter_text:
            self._filtered_items = list(self._items)
        else:
            filter_lower = filter_text.lower()
            self._filtered_items = [
                item
                for item in self._items
                if item.value.lower().startswith(filter_lower)
                or item.label.lower().startswith(filter_lower)
            ]
        # Reset selection when filter changes
        self._selected_index = 0

    def set_selected_index(self, index: int) -> None:
        if not self._filtered_items:
            self._selected_index = 0
        else:
            self._selected_index = _clamp(index, 0, len(self._filtered_items) - 1)

    def invalidate(self) -> None:
        pass

    def render(self, width: int) -> List[str]:
        lines: List[str] = []

        # If no items match filter, show message
        if not self._filtered_items:
            lines.append(self._theme.no_match("  No matching commands"))
            return lines

        primary_column_width = self._get_primary_column_width()

        # Calculate visible range with scrolling
        start_index = max(
            0,
            min(
                self._selected_index - self._max_visible // 2,
                len(self._filtered_items) - self._max_visible,
            ),
        )
        end_index = min(start_index + self._max_visible, len(self._filtered_items))

        # Render visible items
        for i in range(start_index, end_index):
            item = self._filtered_items[i]
            if not item:
                continue

            is_selected = i == self._selected_index
            description_single_line = (
                _normalize_to_single_line(item.description) if item.description else None
            )
            lines.append(
                self._render_item(
                    item, is_selected, width, description_single_line, primary_column_width
                )
            )

        # Add scroll indicators if needed
        if start_index > 0 or end_index < len(self._filtered_items):
            scroll_text = f"  ({self._selected_index + 1}/{len(self._filtered_items)})"
            truncated_scroll = truncate_to_width(scroll_text, width - 2, "")
            lines.append(self._theme.scroll_info(truncated_scroll))

        return lines

    def handle_input(self, key_data: str) -> None:
        kb = get_keybindings()
        # Up arrow - wrap to bottom when at top
        if kb.matches(key_data, "tui.select.up"):
            if self._selected_index == 0:
                self._selected_index = len(self._filtered_items) - 1
            else:
                self._selected_index -= 1
            self._notify_selection_change()
        # Down arrow - wrap to top when at bottom
        elif kb.matches(key_data, "tui.select.down"):
            if self._selected_index == len(self._filtered_items) - 1:
                self._selected_index = 0
            else:
                self._selected_index += 1
            self._notify_selection_change()
        # Enter
        elif kb.matches(key_data, "tui.select.confirm"):
            selected_item = self._get_selected_item()
            if selected_item and self.on_select:
                self.on_select(selected_item)
        # Escape or Ctrl+C
        elif kb.matches(key_data, "tui.select.cancel"):
            if self.on_cancel:
                self.on_cancel()

    def _render_item(
        self,
        item: SelectItem,
        is_selected: bool,
        width: int,
        description_single_line: Optional[str],
        primary_column_width: int,
    ) -> str:
        prefix = "→ " if is_selected else "  "
        prefix_width = visible_width(prefix)

        if description_single_line and width > 40:
            effective_primary_column_width = max(
                1, min(primary_column_width, width - prefix_width - 4)
            )
            max_primary_width = max(1, effective_primary_column_width - PRIMARY_COLUMN_GAP)
            truncated_value = self._truncate_primary(
                item, is_selected, max_primary_width, effective_primary_column_width
            )
            truncated_value_width = visible_width(truncated_value)
            spacing = " " * max(1, effective_primary_column_width - truncated_value_width)
            description_start = prefix_width + truncated_value_width + len(spacing)
            remaining_width = width - description_start - 2  # -2 for safety

            if remaining_width > MIN_DESCRIPTION_WIDTH:
                truncated_desc = truncate_to_width(description_single_line, remaining_width, "")
                if is_selected:
                    return self._theme.selected_text(
                        f"{prefix}{truncated_value}{spacing}{truncated_desc}"
                    )

                desc_text = self._theme.description(spacing + truncated_desc)
                return prefix + truncated_value + desc_text

        max_width = width - prefix_width - 2
        truncated_value = self._truncate_primary(item, is_selected, max_width, max_width)
        if is_selected:
            return self._theme.selected_text(f"{prefix}{truncated_value}")

        return prefix + truncated_value

    def _get_primary_column_bounds(self) -> tuple[int, int]:
        raw_min = (
            self._layout.min_primary_column_width
            or self._layout.max_primary_column_width
            or DEFAULT_PRIMARY_COLUMN_WIDTH
        )
        raw_max = (
            self._layout.max_primary_column_width
            or self._layout.min_primary_column_width
            or DEFAULT_PRIMARY_COLUMN_WIDTH
        )

        return (
            max(1, min(raw_min, raw_max)),
            max(1, max(raw_min, raw_max)),
        )

    def _get_primary_column_width(self) -> int:
        min_width, max_width = self._get_primary_column_bounds()
        widest_primary = 0
        for item in self._filtered_items:
            display_value = self._get_display_value(item)
            widest_primary = max(widest_primary, visible_width(display_value) + PRIMARY_COLUMN_GAP)

        return _clamp(widest_primary, min_width, max_width)

    def _truncate_primary(
        self, item: SelectItem, is_selected: bool, max_width: int, column_width: int
    ) -> str:
        display_value = self._get_display_value(item)
        truncated_value = (
            self._layout.truncate_primary(
                {
                    "text": display_value,
                    "maxWidth": max_width,
                    "columnWidth": column_width,
                    "item": item,
                    "isSelected": is_selected,
                }
            )
            if self._layout.truncate_primary
            else truncate_to_width(display_value, max_width, "")
        )

        return truncate_to_width(truncated_value, max_width, "")

    def _get_display_value(self, item: SelectItem) -> str:
        return item.label or item.value

    def _notify_selection_change(self) -> None:
        selected_item = self._get_selected_item()
        if selected_item and self.on_selection_change:
            self.on_selection_change(selected_item)

    def _get_selected_item(self) -> Optional[SelectItem]:
        if not self._filtered_items:
            return None
        return self._filtered_items[self._selected_index]

    def get_selected_item(self) -> Optional[SelectItem]:
        return self._get_selected_item()

    @property
    def wants_key_release(self) -> bool:
        return False
